src/librustc/mir/repr.rs

```diff
@@ -614,12 +618,60 @@ impl<'tcx> Debug for Lvalue<'tcx> {
 }
 
 ///////////////////////////////////////////////////////////////////////////
+// Scopes
+
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
+pub struct ScopeDataVec {
+    pub vec: Vec<ScopeData>
+}
+
+impl ScopeDataVec {
+    pub fn new() -> Self {
+        ScopeDataVec { vec: Vec::new() }
+    }
+}
+
+impl Index<ScopeId> for ScopeDataVec {
+    type Output = ScopeData;
+
+    #[inline]
+    fn index(&self, index: ScopeId) -> &ScopeData {
+        &self.vec[index.index()]
+    }
+}
+
+impl IndexMut<ScopeId> for ScopeDataVec {
+    #[inline]
+    fn index_mut(&mut self, index: ScopeId) -> &mut ScopeData {
+        &mut self.vec[index.index()]
+    }
+}
+
+#[derive(Copy, Clone, Debug, RustcEncodable, RustcDecodable)]
+pub struct ScopeId(u32);
```

Comment:
At least in the HIR, it would be more practical to use a single generic type IMO.